Add IP Policy guidance (#618)
openjs-foundation/cross-project-council: IP_POLICY_GUIDANCE.md

# OpenJS Foundation IP Policy Guidance

The purpose of this guide is to make it straightforward for projects
to follow the [OpenJS Foundation IP Policy][IP Policy].

Of course, the source of truth remains the [IP Policy][].

If in doubt about anything related to IP after reading this document,
first read the [IP Policy][], and then [ask for help](#getting-help).

## Basic requirements

There are four key requirements for all OpenJS Foundation projects:

1. [Licensing](#1-licensing)
2. [Copyright notices](#2-copyright-notices)
3. Displaying the [standard website footer](#3-standard-website-footer) on the project's website
4. [Adopting the DCO or a CLA](#4-adopting-the-dco-or-a-cla)

### 1. Licensing

Code may be licensed under:
  * [Apache, Version 2.0](http://www.apache.org/licenses/LICENSE-2.0)
  * [MIT](https://opensource.org/licenses/MIT)
  * [2-Clause BSD](https://opensource.org/licenses/BSD-2-Clause)
  * [3-Clause BSD](https://opensource.org/licenses/BSD-3-Clause)

Documentation may be licensed under:
  * [CC BY 4.0](http://creativecommons.org/licenses/by/4.0/)
  * [MIT](https://opensource.org/licenses/MIT)

> If you need to use a license which isn't in the [IP Policy][],
> you will need to obtain permission from the Board to do so.
> Please [open an issue in the CPC repository](https://github.com/openjs-foundation/cross-project-council/issues/new),
> and assign it to @brianwarner.

Author: ```suggestion
If you need to use a license which isn't in the [IP Policy][], you will
[need to obtain permission from the Board to do so](#requiring-an-exemption-from-the-board).
```

# OpenJS Foundation IP Policy Guidance

The purpose of this guide is to make it straightforward for projects
to follow the [OpenJS Foundation IP Policy][IP Policy].

Of course, the source of truth remains the [IP Policy][].

If in doubt about anything related to IP after reading this document,
first read the [IP Policy][], and then [ask for help](#getting-help).

## Basic requirements

There are three key requirements for all OpenJS Foundation projects:

1. [Licensing](#1-licensing),

> 2. [copyright notices](#2-copyright-notices), and

Reviewer: consistent capitalization

3. displaying the [standard website footer](#3-standard-website-footer) on the project's website.

> Additionally, projects may decide to adopt the [OpenJS Foundation Contributor License Agreement (CLA)](#contributor-license-agreement-cla).

Reviewer 2: Should this not actually be a fourth basic requirement. `Incoming code must be made under a DCO or optionally an CLA` instead?

Reviewer 2: See that this is what @kborchers pointed out.

Reviewer 3: I've got a patch coming that does this, yes :-)

Author: Yup. This has been clarified.

### 1. Licensing

Code may be licensed under:
  * [Apache, Version 2.0](http://www.apache.org/licenses/LICENSE-2.0)
  * [MIT](https://opensource.org/licenses/MIT)
  * [2-Clause BSD](https://opensource.org/licenses/BSD-2-Clause)
  * [3-Clause BSD](https://opensource.org/licenses/BSD-3-Clause)

Documentation may be licensed under:
  * [CC BY 4.0](http://creativecommons.org/licenses/by/4.0/)
  * [MIT](https://opensource.org/licenses/MIT)

If you need to use a license which isn't in the list,

> you will need to obtain permission from the Board to do so.

Reviewer: how does a project go about this?

Reviewer 3: ```suggestion
you will need to obtain permission from the Board to do so. Please [open an issue in the CPC repo](https://github.com/openjs-foundation/cross-project-council/issues), and assign it to brianwarner.
```

### 2. Copyright notices

The recommendation is to use a general copyright statement
of the following form (where XYZ is the project's name):

- **Copyright The XYZ Authors.**
- **Copyright The XYZ Contributors.**
- **Copyright Contributors to the XYZ project.**

By using this format, the project avoids having to deal with
names of copyright holders, years or ranges of years,
and variations on the (c) symbol.

_Please note that you **must not** change or remove existing copyright lines
unless you put them there **and** have the right to do so.

> If there are existing copyright lines just

Reviewer: remove "just". maybe wants a comma after "lines"

Author: ```suggestion
If there are existing copyright lines
```

add the recommended copyright statement below the existing copyright lines.

> You may however update copyright lines that say

Reviewer: must?

Reviewer 3: Yes, I'd recommend keeping this as "must." While these lines are informational in nature, they were likely put there for a specific purpose and a number of licenses require these to be distributed intact (unless, of course, the person who holds the copyright removes the line themselves).

Author: That comment isn't clear to me.

@boneskull are you suggesting the following change?

```suggestion
You must however update copyright lines that say
```

@brianwarner is that what you understood?

If that's the proposed change, I'd be opposed: this can create a lot of work for certain projects for no additional benefits but consistency.

Reviewer 2: @tobie my read of @brianwarner's comment at least is that it applies to `_Please note that you **must not** change or remove existing copyright lines`, not sure about @boneskull's comment.  Unless its a hard requirement to update the copyright lines to reflect the OpenJS foundation I'd agree that `may` in that context makes the most sense.

Reviewer 3: Sorry that was unclear, I was referring back to the last use of "must". 

Agreed, "may" is best here. The idea is that you can update any lines where you have the right, but you don't have to do so. At best, these lines are hints as to copyright ownership, and it would be asking projects to do busy-work to update them.

Author: @boneskull are you OK with not changing this?

Author: Pinging @boneskull, here.

Reviewer: yes

Author: Awesome. Thank you.

"JS Foundation" or "Node.js Foundation" to "OpenJS Foundation"
as this simply reflects the merged status of the organizations._

### 3. Standard website footer

All project websites need to have a [standard website footer][].

### Contributor License Agreement (CLA)

> Projects may decide to adopt the [OpenJS Foundation CLA][CLA] but may not use other CLAs.

Reviewer: add DCO (and link) as mentioned by @kborchers

Reviewer 3: I'll send a patch on this one, since it involved changes in multiple parts of the file. I'll also add the link to https://developercertificate.org.

- **Copyright The XYZ Authors.**
- **Copyright The XYZ Contributors.**
- **Copyright Contributors to the XYZ project.**

By using this format, the project avoids having to deal with
names of copyright holders, years or ranges of years,
and variations on the (c) symbol.

_Please note that you **must not** change or remove existing copyright lines
unless you put them there **and** have the right to do so.
If there are existing copyright lines
add the recommended copyright statement below the existing copyright lines.
You may however update copyright lines that say
"JS Foundation" or "Node.js Foundation" to "OpenJS Foundation"
as this simply reflects the merged status of the organizations._

### 3. Standard website footer

All project websites need to have a [standard website footer][].

### 4. Contributor License Agreement (CLA)

By default the OpenJS Foundation uses the [DCO][], which means contributors
submit their contributions under the project's designated open source license.
Every commit must contain a Signed-off-by line in the commit message (e.g.,
`Signed-off-by: Authorname <author@email.com>`), indicating the author has the

> right to make the contribution under the open source license.

Author: This feels like an implementation detail of the DCO that begs more questions:

- What happens when there are multiple authors?
- What happens when commits are squashed?
- Can I sign off for someone else when I merge a commit?
- What if I'm batch-merging commits made with GitHub's "insert a suggestion" feature?
- Is a Sign-off-by in the body of the pull request good enough?
- Etc.

My suggestion would be to have a separate documents with implementation details and best practices for implementing the DCO and CLA and dealing with them on a day to day basis.

Reviewer 3: In general those others are definitely implementation details and would be great in a breakout, but I strongly recommend keeping this as it explains what the DCO is and how it differs from the next option. I suspect many don't make the association between a signed-off-by line and the DCO (much as many don't realize that there's only one DCO).

Author: OK, so I tried to include a small description of what needs to happen from a maintainer's perspective to implement the DCO. I have more questions: Does the project need to mention the DCO anywhere? What's language should it use? Should it set up infra? Etc.

Reviewer 3: That's a good point, I think it would be worth me putting together some guidelines on this. I'll take this one on.

# OpenJS Foundation IP Policy Guidance

The purpose of this guide is to make it straightforward for projects
to follow the [OpenJS Foundation IP Policy][IP Policy].

Of course, the source of truth remains the [IP Policy][].

If in doubt about anything related to IP after reading this document,
first read the [IP Policy][], and then [ask for help](#getting-help).

## Basic requirements

There are four key requirements for all OpenJS Foundation projects:

1. [Licensing](#1-licensing)
2. [Copyright notices](#2-copyright-notices)
3. Displaying the [standard website footer](#3-standard-website-footer) on the project's website
4. [Adopting the DCO or a CLA](#4-adopting-the-dco-or-a-cla)

### 1. Licensing

Code may be licensed under:
  * [Apache, Version 2.0](http://www.apache.org/licenses/LICENSE-2.0)
  * [MIT](https://opensource.org/licenses/MIT)
  * [2-Clause BSD](https://opensource.org/licenses/BSD-2-Clause)
  * [3-Clause BSD](https://opensource.org/licenses/BSD-3-Clause)

Documentation may be licensed under:
  * [CC BY 4.0](http://creativecommons.org/licenses/by/4.0/)
  * [MIT](https://opensource.org/licenses/MIT)

> If you need to use a license which isn't in the [IP Policy],

Reviewer 4: I think this needs `[]` after `[IP Policy]`

Author: ```suggestion
If you need to use a license which isn't in the [IP Policy][],
```

you will need to obtain permission from the Board to do so.

> Please [open an issue in the CPC repository][new cpc issue],

Reviewer 4: Where is `[new cpc issue]:` defined?

Do we need

    [new cpc issue]: https://github.com/openjs-foundation/cross-project-council/issues/new

at the bottom?

Author: ```suggestion
Please [open an issue in the CPC repository](https://github.com/openjs-foundation/cross-project-council/issues/new),
```

and assign it to @brianwarner.

### 2. Copyright notices

The recommendation is to use a general copyright statement
of the following form (where XYZ is the project's name):

- **Copyright The XYZ Authors.**
- **Copyright The XYZ Contributors.**
- **Copyright Contributors to the XYZ project.**

By using this format, the project avoids having to deal with
names of copyright holders, years or ranges of years,
and variations on the (c) symbol.

_Please note that you **must not** change or remove existing copyright lines
unless you put them there **and** have the right to do so.
If there are existing copyright lines
add the recommended copyright statement below the existing copyright lines.
You may however update copyright lines that say
"JS Foundation" or "Node.js Foundation" to "OpenJS Foundation"
as this simply reflects the merged status of the organizations._

### 3. Standard website footer

All project websites need to have a [standard website footer][].

### 4. Adopting the DCO or a CLA

> Each project much adopt either the [Developer Certificate of Origin (DCO)][DCO]

Reviewer 3: ```suggestion
Each project must adopt either the [Developer Certificate of Origin (DCO)][DCO]
```

or a Contributor License Agreement (CLA).

The Board has pre-approved an [Individual CLA (ICLA)][ICLA]
and a [Corporate CLA (CCLA)][CCLA]
which projects can adopt without any further review.

If you need to use a different CLA, you will need to
[require an exemption from the board](#requiring-an-exemption-from-the-board).

> To setup the CLA infrastructure, [reach out to the foundation staff](#getting-help).

Author: @brianwarner: Does the foundation also offer DCO infra?

Reviewer 3: @tobie We don't, as the DCO process is completely contained within the git log. We typically recommend projects use Probot DCO as it makes it easier for maintainers to verify that the signoff exists, but that's just for convenience.

> To setup the CLA infrastructure, [reach out to the foundation staff](#getting-help).

Author: ```suggestion
We are preparing additional guidance documentation on how to implement the DCO or a CLA,
what infrastructure the foundation provides to help with this,
what steps maintainers need to make when merging pull requests,
and how to fix common issues that arise when using either solution.

To setup the CLA infrastructure until this additional documentation is ready,
please [reach out to foundation staff](#getting-help).
```

Please [open an issue in the CPC repository][new cpc issue],
and assign it to @brianwarner.

### 2. Copyright notices

The recommendation is to use a general copyright statement
of the following form (where XYZ is the project's name):

- **Copyright The XYZ Authors.**
- **Copyright The XYZ Contributors.**
- **Copyright Contributors to the XYZ project.**

By using this format, the project avoids having to deal with
names of copyright holders, years or ranges of years,
and variations on the (c) symbol.

_Please note that you **must not** change or remove existing copyright lines
unless you put them there **and** have the right to do so.
If there are existing copyright lines
add the recommended copyright statement below the existing copyright lines.
You may however update copyright lines that say
"JS Foundation" or "Node.js Foundation" to "OpenJS Foundation"
as this simply reflects the merged status of the organizations._

### 3. Standard website footer

All project websites need to have a [standard website footer][].

### 4. Adopting the DCO or a CLA

Each project much adopt either the [Developer Certificate of Origin (DCO)][DCO]

> or Contributor License Agreement (CLA).

Author: ```suggestion
or a Contributor License Agreement (CLA).
```

If you need to use a license which isn't in the [IP Policy],
you will need to obtain permission from the Board to do so.
Please [open an issue in the CPC repository][new cpc issue],
and assign it to @brianwarner.

### 2. Copyright notices

The recommendation is to use a general copyright statement
of the following form (where XYZ is the project's name):

- **Copyright The XYZ Authors.**
- **Copyright The XYZ Contributors.**
- **Copyright Contributors to the XYZ project.**

By using this format, the project avoids having to deal with
names of copyright holders, years or ranges of years,
and variations on the (c) symbol.

_Please note that you **must not** change or remove existing copyright lines
unless you put them there **and** have the right to do so.
If there are existing copyright lines
add the recommended copyright statement below the existing copyright lines.
You may however update copyright lines that say
"JS Foundation" or "Node.js Foundation" to "OpenJS Foundation"
as this simply reflects the merged status of the organizations._

### 3. Standard website footer

All project websites need to have a [standard website footer][].

### 4. Adopting the DCO or a CLA

> Each project much adopt either the [Developer Certificate of Origin (DCO)](DCO)

Reviewer 5: resolves the lint error 
```suggestion
Each project much adopt either the [Developer Certificate of Origin (DCO)][DCO]
```

Reviewer 6: 106215

  * [MIT](https://opensource.org/licenses/MIT)

If you need to use a license which isn't in the [IP Policy][], you will
[need to obtain permission from the Board to do so](#requiring-an-exemption-from-the-board).

### 2. Copyright notices

The recommendation is to use a general copyright statement
of the following form (where XYZ is the project's name):

- **Copyright The XYZ Authors.**
- **Copyright The XYZ Contributors.**
- **Copyright Contributors to the XYZ project.**

By using this format, the project avoids having to deal with
names of copyright holders, years or ranges of years,
and variations on the (c) symbol.

_Please note that you **must not** change or remove existing copyright lines
unless you put them there **and** have the right to do so.
If there are existing copyright lines
add the recommended copyright statement below the existing copyright lines.
You may however update copyright lines that say
"JS Foundation" or "Node.js Foundation" to "OpenJS Foundation"
as this simply reflects the merged status of the organizations._

### 3. Standard website footer

All project websites need to have a [standard website footer][].

### 4. Adopting the DCO or a CLA

> Each project must adopt either the [Developer Certificate of Origin (DCO)][DCO]

Reviewer 7: I disagree that this should be the policy, but this PR isn’t imposing it, merely documenting it, so i won’t derail on that here :-)

Author: Yeah, thanks for your understanding.

or a Contributor License Agreement (CLA).

The Board has pre-approved an [Individual CLA (ICLA)][ICLA]
and a [Corporate CLA (CCLA)][CCLA]
which projects can adopt without any further review.

If you need to use a different CLA, you will need to

> [require an exemption from the board](#requiring-an-exemption-from-the-board).

Reviewer 8: ```suggestion
[obtaining an exemption from the board](#obtaining-an-exemption-from-the-board).
```

Author: Thank you for this fix. I think I wanted to say "request" and ended-up using "require" instead. Is "request" or "obtain" better here?

Author: ^ @kborchers

Reviewer 7: probably obtain, since that implies request, but the reverse isn’t true

Reviewer 8: Yeah, I would still say obtain

We are preparing additional guidance documentation on how to implement the DCO or a CLA,
what infrastructure the foundation provides to help with this,
what steps maintainers need to make when merging pull requests,
and how to fix common issues that arise when using either solution.

To setup the CLA infrastructure until this additional documentation is ready,
please [reach out to foundation staff](#getting-help).

> ## Requiring an exemption from the Board

Reviewer 8: ```suggestion
## Obtaining an exemption from the Board
```
